src/lib.rs

```diff
@@ -463,41 +452,289 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> HotShot<TYPES::ConsensusType
+    async fn spawn_all(&self) -> HotShotHandle<TYPES, I>;
+
+    /// decide which handler to call based on the message variant and `transmit_type`
+    async fn handle_message(&self, item: Message<TYPES, I>, transmit_type: TransmitType) {
+        match (item.kind, transmit_type) {
+            (MessageKind::Consensus(msg), TransmitType::Broadcast) => {
+                self.handle_broadcast_consensus_message(msg, item.sender)
+                    .await;
+            }
+            (MessageKind::Consensus(msg), TransmitType::Direct) => {
+                self.handle_direct_consensus_message(msg, item.sender).await;
+            }
+            (MessageKind::Data(msg), TransmitType::Broadcast) => {
+                self.handle_broadcast_data_message(msg, item.sender).await;
+            }
+            (MessageKind::Data(msg), TransmitType::Direct) => {
+                self.handle_direct_data_message(msg, item.sender).await;
+            }
+            (MessageKind::_Unreachable(_), _) => unimplemented!(),
+        };
+    }
+
     /// Handle an incoming [`ConsensusMessage`] that was broadcasted on the network.
+    async fn handle_broadcast_consensus_message(
+        &self,
+        msg: I::ConsensusMessage,
+        sender: TYPES::SignatureKey,
+    );
+
+    /// Handle an incoming [`ConsensusMessage`] directed at this node.
+    async fn handle_direct_consensus_message(
+        &self,
+        msg: I::ConsensusMessage,
+        sender: TYPES::SignatureKey,
+    );
+
+    /// Handle an incoming [`DataMessage`] that was broadcasted on the network
+    async fn handle_broadcast_data_message(
+        &self,
+        msg: DataMessage<TYPES>,
+        _sender: TYPES::SignatureKey,
+    ) {
+        // TODO validate incoming broadcast message based on sender signature key
+        match msg {
+            DataMessage::SubmitTransaction(transaction, _view_number) => {
+                let size = bincode_opts().serialized_size(&transaction).unwrap_or(0);
+
+                // The API contract requires the hash to be unique
+                // so we can assume entry == incoming txn
+                // even if eq not satisfied
+                // so insert is an idempotent operation
+                let mut new = false;
+                self.transactions()
+                    .modify(|txns| {
+                        new = txns.insert(transaction.commit(), transaction).is_none();
+                    })
+                    .await;
+
+                if new {
+                    // If this is a new transaction, update metrics.
+                    let consensus = self.hotstuff().read().await;
+                    consensus.metrics.outstanding_transactions.update(1);
+                    consensus
+                        .metrics
+                        .outstanding_transactions_memory_size
+                        .update(i64::try_from(size).unwrap_or(i64::MAX));
```

Comment:
Should this warn if the conversion fails? Might be useful in the case of overflow.